From 6872cc8e705ee5a9080dcded4827904cba235085 Mon Sep 17 00:00:00 2001
From: Super User <root@localhost.localdomain>
Date: Sat, 16 May 2026 11:53:07 +0800
Subject: [PATCH 2/2] KVM: arm64: Add debug logging for HDBSS MMIO tracking

Only print when DBM would be incorrectly added to MMIO.
This avoids log flood and focuses on the real bug.

Signed-off-by: OpenEuler Kernel Team <kernel@openeuler.org>
---
 arch/arm64/kvm/mmu.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/arch/arm64/kvm/mmu.c b/arch/arm64/kvm/mmu.c
index 34061605e906..d529ffe87a7f 100644
--- a/arch/arm64/kvm/mmu.c
+++ b/arch/arm64/kvm/mmu.c
@@ -1682,6 +1682,13 @@ static int user_mem_abort(struct kvm_vcpu *vcpu, phys_addr_t fault_ipa,
 	if (writable)
 		prot |= KVM_PGTABLE_PROT_W;
 
+	/* Debug: detect if DBM would be incorrectly added to MMIO */
+	if (device && writable && kvm->arch.enable_hdbss && logging_active) {
+		pr_info_ratelimited("KVM_HDBSS_DBG: BUG detected! "
+			"DBM would be added to MMIO ipa=0x%llx dev=%d en=%d log=%d\n",
+			fault_ipa, device, kvm->arch.enable_hdbss, logging_active);
+	}
+
 	if (!device && writable && kvm->arch.enable_hdbss && logging_active)
 		prot |= KVM_PGTABLE_PROT_DBM;
 
-- 
2.43.0

